File: blocklib/grident/test/test_tx_control.cc
#include <gnuradio/grident/tx_control.h>

#include <cassert>
#include <cstring>

using gr::grident::parse_tx_control_message;
using gr::grident::tx_control_state;

int main()
{
    assert(parse_tx_control_message("PTT_ON", 6) == tx_control_state::on);
    assert(parse_tx_control_message("ptt_off", 7) == tx_control_state::off);
    assert(parse_tx_control_message("TX", 2) == tx_control_state::on);
    assert(parse_tx_control_message("RX", 2) == tx_control_state::off);
    assert(parse_tx_control_message("1", 1) == tx_control_state::on);
    assert(parse_tx_control_message("0", 1) == tx_control_state::off);
    assert(parse_tx_control_message(R"({"ptt":true})", 12) == tx_control_state::on);
    assert(parse_tx_control_message(R"({"ptt":false})", 13) == tx_control_state::off);
    assert(parse_tx_control_message(R"({"tx":1})", 8) == tx_control_state::on);
    assert(parse_tx_control_message("UNKNOWN", 7) == tx_control_state::unknown);

    static constexpr unsigned char k_sot[] = { 0x02, 0x00, 0x03, 'S', 'O', 'T' };
    static constexpr unsigned char k_eot[] = { 0x02, 0x00, 0x03, 'E', 'O', 'T' };
    assert(parse_tx_control_message(reinterpret_cast<const char*>(k_sot), sizeof(k_sot)) == tx_control_state::on);
    assert(parse_tx_control_message(reinterpret_cast<const char*>(k_eot), sizeof(k_eot)) == tx_control_state::off);
    assert(parse_tx_control_message("SOT", 3) == tx_control_state::on);
    assert(parse_tx_control_message("EOT", 3) == tx_control_state::off);
    return 0;
}

File: python/tests/test_tx_control.py
import unittest

from grident.tx_control import (
    TxControlState,
    format_linht_pmt,
    format_tx_control_message,
    parse_linht_pmt,
    parse_tx_control_message,
)


class TestTxControl(unittest.TestCase):
    def test_plain_on_off(self):
        self.assertEqual(parse_tx_control_message(b"PTT_ON"), TxControlState.ON)
        self.assertEqual(parse_tx_control_message("ptt_off"), TxControlState.OFF)
        self.assertEqual(parse_tx_control_message("TX"), TxControlState.ON)
        self.assertEqual(parse_tx_control_message("RX"), TxControlState.OFF)

    def test_json(self):
        self.assertEqual(parse_tx_control_message('{"ptt": true}'), TxControlState.ON)
        self.assertEqual(parse_tx_control_message('{"ptt": false}'), TxControlState.OFF)
        self.assertEqual(parse_tx_control_message('{"tx": 0}'), TxControlState.OFF)

    def test_format_roundtrip(self):
        for state in TxControlState:
            payload = format_tx_control_message(state, as_json=True)
            self.assertEqual(parse_tx_control_message(payload), state)

    def test_linht_pmt(self):
        self.assertEqual(parse_linht_pmt(format_linht_pmt("SOT")), TxControlState.ON)
        self.assertEqual(parse_linht_pmt(format_linht_pmt("EOT")), TxControlState.OFF)
        self.assertEqual(parse_tx_control_message(format_linht_pmt("SOT")), TxControlState.ON)
        self.assertEqual(parse_tx_control_message(format_linht_pmt("EOT")), TxControlState.OFF)
        self.assertEqual(parse_tx_control_message("SOT"), TxControlState.ON)
        self.assertEqual(parse_tx_control_message("EOT"), TxControlState.OFF)


if __name__ == "__main__":
    unittest.main()
